salt/metalk8s/orchestrate/upgrade/init.sls

```diff
@@ -0,0 +1,82 @@
+{%- set dest_version = pillar.orchestrate.dest_version %}
+
+Execute the upgrade prechecks:
+  salt.runner:
+    - name: state.orchestrate
+    - mods:
+      - metalk8s.orchestrate.upgrade.precheck
+    - saltenv: {{ saltenv }}
+    - pillar:
+        orchestrate:
+          dest_version: {{ dest_version }}
+
+Upgrade etcd cluster:
+  salt.runner:
+    - name: state.orchestrate
+    - mods:
+      - metalk8s.orchestrate.upgrade.etcd
+    - saltenv: {{ saltenv }}
+    - pillar:
+        orchestrate:
+          dest_version: {{ dest_version }}
+    - require:
+      - salt: Execute the upgrade prechecks
+
+{%- set cp_nodes = salt.metalk8s.minions_by_role('master') | sort %}
+{%- set other_nodes = pillar.metalk8s.nodes.keys() | difference(cp_nodes) | sort %}
+
+{%- for node in cp_nodes + other_nodes %}
+
+  {%- set kubeconfig = "/etc/kubernetes/admin.conf" %}
+  {%- set context = "kubernetes-admin@kubernetes" %}
+
+  {%- set node_version = pillar.metalk8s.nodes[node].version %}
+  {#- 1 = Upgrade, -1 = Downgrade, 0 = Correct version #}
+  {%- set upgrade = salt.pkg.version_cmp(dest_version, node_version) %}
+
+Set node {{ node }} version to {{ dest_version }}:
+  metalk8s_kubernetes.node_label_present:
+    - name: metalk8s.scality.com/version
+    - node: {{ node }}
+    - value: "{{ dest_version }}"
+    - kubeconfig: {{ kubeconfig }}
+    - context: {{ context }}
+    - require:
+      - salt: Upgrade etcd cluster
+  {%- if previous_node is defined %}
+      - salt: Deploy node {{ previous_node }}
+  {%- endif %}
+
+Deploy node {{ node }}:
+  salt.runner:
+    - name: state.orchestrate
+    - mods:
+      - metalk8s.orchestrate.deploy_node
+  {#- Always use the newest saltenv between dest_version and current_version,
+      Upgrade use dest_version, Downgrade use the current version #}
+  {%- if upgrade == 1 %}
+    - saltenv: metalk8s-{{ dest_version }}
```

Comment:
No, I don't think this is how it should work.

We have three versions at play:

- The current version/saltenv currently deployed on the minion. We don't really care about that one.
- The target version deployed on the minion. That one is set in the call above, good.
- The version of MetalK8s *used to deploy/upgrade/... the minion*, i.e. the version of the 'cluster', the version of the SLSs used *on the master*. This should *not* be dependent on the (target) version of the minion

So, this is `saltenv: {{ saltenv }}` like any other.

Reply:
Hmm indeed, btw I did a check in precheck to check that the saltenv is the right so yes we can use the saltenv directly but that should be the same as this